dill/_dill.py

```diff
@@ -876,12 +939,89 @@ def __getattribute__(self, attr):
             attrs[index] = ".".join([attrs[index], attr])
         return type(self)(attrs, index)
 
+
+
+# Used to stay compatible with versions of dill whose _create_cell functions
+# do not have a default value.
+# Can be safely replaced removed entirely (replaced by empty tuples for calls to
+# _create_cell) once breaking changes are allowed.
+_CELL_REF_shim = None
+
 if PY3:
-    def _create_cell(contents):
-        return (lambda y: contents).__closure__[0]
+    def _create_cell(contents=None):
+        return (lambda: contents).__closure__[0]
+
+    if OLD37:
+        # Python 3.0 to 3.6 is in a weird case, where it is possible to pickle
+        # recursive cells, we can't assign directly to the cell.
+        # https://stackoverflow.com/a/59276835
+        __nonlocal = ('nonlocal cell',) if PY3 else ('',)
+        exec('''def _setattr(cell, name, value):
+            if type(cell) is CellType and name == 'cell_contents':
+                def cell_setter(value):
+                    %s
+                    cell = value # pylint: disable=unused-variable
+                func = FunctionType(cell_setter.__code__, globals(), "", None, (cell,)) # same as cell_setter, but with cell being the cell's contents
+                func(value)
+            else:
+                setattr(cell, name, value)''' % __nonlocal)
+
+        exec('''def _delattr(cell, name):
+            if type(cell) is CellType and name == 'cell_contents':
+                def cell_deleter():
+                    %s
+                    del cell # pylint: disable=unused-variable
+                func = FunctionType(cell_deleter.__code__, globals(), "", None, (cell,)) # same as cell_deleter, but with cell being the cell's contents
+                func()
+            else:
+                delattr(cell, name)''' % __nonlocal)
+
 else:
-    def _create_cell(contents):
-        return (lambda y: contents).func_closure[0]
+    def _create_cell(contents=None):
+        return (lambda: contents).func_closure[0]
+
+    if IS_PYPY:
+        from . import nonlocals as _nonlocals
+        @_nonlocals.export_nonlocals('cellv')
+        def _setattr(cell, name, value):
+            if type(cell) is CellType and name == 'cell_contents':
+                cellv = None
+                @_nonlocals.nonlocals('cellv', closure_override=(cell,))
+                def cell_setter(value):
+                    cellv = value # pylint: disable=unused-variable
+                cell_setter(value)
+            else:
+                setattr(cell, name, value)
+
+        def _delattr(object, name):
+            if type(object) is CellType and name == 'cell_contents':
+                raise NotImplementedError('Empty cells in PyPy2')
+            else:
+                delattr(object, name)
+
+    else: # CPython 2
+        def _create_cell(contents=None):
+            return (lambda: contents).func_closure[0]
+
+        _PyCell_Set = ctypes.pythonapi.PyCell_Set
```

Comment:
Needs HAS_CTYPES guard before this line (for versions of python that ship without ctypes)

Reply:
I think that `not IS_PYPY` is sufficient because `IS_PYPY` implies `not HAS_CTYPES`.

Reply:
No.  CPython ships without ctypes and other modules in certain instances.  So, it needs the `HAS_CTYPES` guard.  Some pre-built cpython, including google app engine, don't build `ctypes`.